File: Lecture18/na2cl2.py
from nacl import *
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

name="Na2Cl2"       # for output files
nNa = 2
nCl = 2
n = nNa + nCl
a = 0.2
r_Na = [  [ 0, 0, 0 ], [ a, a, 0] ]
r_Cl = [  [ a, 0, 0 ], [ 0, a, 0 ] ]

# Initialize the cluster, add guesses at the
# minimum arrangement. 
cluster = Cluster()

for i in xrange(nNa) :
    r = Vector(r_Na[i])
    cluster.add(Na(r))

for i in xrange(nCl) :
    r = Vector(r_Cl[i])
    cluster.add(Cl(r))

print " " + name + " cluster"
print " Initial potential energy = " + str( cluster.potential_energy() )

# Minimize the function
accuracy = 1e-6

res = cluster.minimize( accuracy )

pe = res[1]
iterations = res[4]

# Print out resulting files, and also
# plot the values in matplotlib
print " Minimized potential energy = " + str(pe) + " eV"
print " Binding energy of cluster  = " + str( pe / 2.0 ) + " eV"
print " Number of function calls = " + str( iterations )

file_name = name + ".data"
outfile = open( file_name, 'w' )
for i in xrange( nNa + nCl - 1) :
    for j in range(i+1,nNa+nCl) :
        rij = cluster.ion(i).r - cluster.ion(j).r
        dr = sqrt( np.dot(rij,rij) )
        s =  "(" + cluster.ion(i).name + ")-(" + cluster.ion(j).name + ")"
        print " " + s + " r_" + str(i) + str(j) + " = " + str( dr ) + " nm"

outfile.write( str(cluster) )
outfile.close()

fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

[x,y,z] = cluster.convert()
ax.scatter( x,y,z )
plt.show()

File: Lecture18/na2cl2.cpp
#include "nacl.hpp"

int main()
{
    string name("Na2Cl2");      // for output files
    const int nNa = 2, nCl = 2;
    const int n = nNa + nCl;
    double a = 0.2;
    double r_Na[nNa][3] = { { 0, 0, 0 }, { a, a, 0} };
    double r_Cl[nCl][3] = { { a, 0, 0 }, { 0, a, 0 } };

    Cluster cluster;

    for (int i = 0; i < nNa; i++) {
        Matrix<double,1> r(3, r_Na[i]);
        cluster.add(Na(r));
    }
    for (int i = 0; i < nCl; i++) {
        Matrix<double,1> r(3, r_Cl[i]);
        cluster.add(Cl(r));
    }
    cout << " " << name << " cluster" << endl
         << " Initial potential energy = "
         << cluster.potential_energy() << endl;

    int iterations = 0;
    double accuracy = 1e-6;

    double pe = cluster.minimize( accuracy, iterations );


    cout << " Minimized potential energy = " << pe << "\n"
         << " Binding energy of cluster  = " << pe / 2 << " eV\n"
         << " Number of iterations = " << iterations << endl;

    string file_name = name + ".data";
    ofstream file(file_name.c_str());
    for (int i = 0; i < nNa + nCl - 1; i++) {
        for (int j = i + 1; j < nNa + nCl; j++) {
	    Matrix<double,1> rij = cluster.ion(i).r - cluster.ion(j).r;
            double dr = sqrt(dot(rij, rij));
            string s = "(" + cluster.ion(i).name + ")-(" +
	      cluster.ion(j).name + ")";
            cout << " " << s << " r_" << i << j << " = " << dr << " nm\n";
        }
    }
    file << cluster;
    file.close();
    cout << " xyz coordinates in " << file_name << endl;

    file_name = name + ".gnu";
    file.open(file_name.c_str());
    file << "set title \"(NaCl)_2 Energy = " << pe / 2 << " eV\"\n"
         << "set hidden3d\n"
         << "unset key\n"
         << "splot \"na2cl2.data\" with points lt 3 pt 7 ps 8\n"
         << "pause -1 \"Hit return to continue \"\n";
    file.close();
    cout << " Gnuplot script in " << file_name << endl;
}
